src/Evolution/Systems/CurvedScalarWave/Worldtube/PunctureField.cpp

```diff
@@ -0,0 +1,442 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/CurvedScalarWave/Worldtube/PunctureField.hpp"
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/DynamicBuffer.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace CurvedScalarWave::Worldtube {
+
+void compute_puncture_field(
+    gsl::not_null<Variables<tmpl::list<
+        Tags::Psi, ::Tags::dt<Tags::Psi>,
+        ::Tags::deriv<Tags::Psi, tmpl::size_t<3>, Frame::Inertial>>>*>
+        evolved_vars,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& coords, const double time,
+    const double orbital_radius, const double bh_mass, const size_t order) {
+  if (order == 0) {
+    compute_puncture_field_0(evolved_vars, coords, time, orbital_radius,
+                             bh_mass);
+  } else {
+    ERROR(
+        "Puncture order is only implemented up to order 0 but you requested "
+        "order "
+        << order);
+  }
+}
+
+void compute_puncture_field_0(
+    gsl::not_null<Variables<tmpl::list<
+        Tags::Psi, ::Tags::dt<Tags::Psi>,
+        ::Tags::deriv<Tags::Psi, tmpl::size_t<3>, Frame::Inertial>>>*>
+        evolved_vars,
```

Comment:
Question @nilsdeppe: do you have any preference these days on passing Variables vs individual Tensors to functions like this? I often found it easier to work exclusively with Tensors in pointwise functions because then the function is agnostic to where the Tensors are actually stored (could be in separate Variables or in one Variables together with other things, or on a Variables with prefixes tags). Now we have nonowning variables so we could create suitable Variables when calling this function. But overall my experience that avoiding Variables and Tags in pointwise functions makes everything easier. Thoughts?

Reply:
A quick reply now, I can elaborate more if you want. I've started adding Variables-based overloads to GRMHD and GH for things like time derivatives. The motivation there being that 1. there are fewer arguments 2. we can now use structured bindings to easily extract out of a Variables 3. non-owning Variables makes it easier to deal with and 4. the `DataStructures/TaggedContainer.hpp` code makes working with multiple tagged containers simultaneously very easy and so overcomes another hurdle we had with using Variables more.

Basically, I think there are cases in generic code where manipulating a Variables (or set of them) is significantly easier than trying to get the right Tensor from the right tagged container, and so I'm generally in favor of doing so when it make the code a lot easier to read and write. E.g. there were places in GHGRMHD where we would take a bunch of takes from 3 or so tagged containers, put references of them in a TuppleTuple, just so we could pull them out later. This was extremely parameter pack and template dense code that made it hard to read, but the recent additions I listed above make it so that isn't needed anymore.

Reply:
Sounds reasonable, I'll give the new tools a try. Maybe generic Variables would also be a way to add new evolution variables like recently the electron fraction without refactoring half the code base.